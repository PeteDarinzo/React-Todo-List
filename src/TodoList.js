import React, { useState } from "react";
import Todo from "./Todo"
import NewTodoForm from "./NewTodoForm";
import { v4 as uuid } from "uuid";
import "./ToDoList.css"



const TodoList = () => {

  const [tasks, setTasks] = useState([]);

  // set tasks from local storage
  window.addEventListener('DOMContentLoaded', (e) => {
    if (localStorage.getItem("tasks") === null) {
      localStorage.setItem("tasks", JSON.stringify(tasks));
    }
    setTasks(() => JSON.parse(localStorage.getItem("tasks")));
  });

  const addTask = (newTask) => {
    setTasks((tasks) => {
      const newTasks = [...tasks, { ...newTask, id: uuid() }];
      localStorage.setItem("tasks", JSON.stringify(newTasks));
      return newTasks;
    });
  }

  const toggleComplete = (id) => {
    let task = {};
    for (let t of tasks) {
      if (t.id === id) {
        task = t;
      }
    }
    deleteTask(id);
    task.completed === false ? task.completed = true : task.completed = false;
    setTasks((tasks) => {
      const newTasks = [...tasks, { ...task }];
      localStorage.setItem("tasks", JSON.stringify(newTasks));
      return newTasks
    });
  }

  const deleteTask = (id) => {
    setTasks((tasks) => {
      const newTasks = tasks.filter(task => task.id !== id);
      localStorage.setItem("tasks", JSON.stringify(newTasks));
      return newTasks;
    });
  }

  return (
    <div>
      <h2>TODO List</h2>
      <NewTodoForm addTask={addTask} />
      <div className="ToDoList">
        {tasks.map(({ task, completed, id }) => <Todo task={task} completed={completed} toggleComplete={toggleComplete} deleteTask={deleteTask} key={id} id={id} />)}
      </div>
    </div>
  );
}

export default TodoList;