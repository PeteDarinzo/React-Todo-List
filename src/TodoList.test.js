import React from "react";
import { render, fireEvent } from '@testing-library/react';
import TodoList from './TodoList';



it("renders a todo list without crashing", () => {
  <TodoList />
});

it("matches the todo snapshot", () => {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it("should add a new task", () => {
  const { queryByText, getByLabelText } = render(<TodoList />);

  const taskInput = getByLabelText("New Task");
  const btn = queryByText("Submit")

  fireEvent.change(taskInput, { target: { value: "walk dog" } });
  fireEvent.click(btn);

  const newTask = queryByText("walk dog");
  
  expect(newTask).toBeInTheDocument();
});

it("should delete a task", () => {
  const { queryByText, getByLabelText } = render(<TodoList />);

  const taskInput = getByLabelText("New Task");
  const addBtn = queryByText("Submit")

  fireEvent.change(taskInput, { target: { value: "feed cat" } });
  fireEvent.click(addBtn);

  const newTask = queryByText("feed cat");

  expect(newTask).toBeInTheDocument();

  const deleteBtn = queryByText("X");

  fireEvent.click(deleteBtn);

  expect(newTask).not.toBeInTheDocument();
});

it("should toggle a task's completion", () => {
  const { queryByText, getByLabelText } = render(<TodoList />);

  const taskInput = getByLabelText("New Task");
  const addBtn = queryByText("Submit")

  fireEvent.change(taskInput, { target: { value: "feed turtle" } });
  fireEvent.click(addBtn);

  const newTask = queryByText("feed turtle");

  expect(newTask).toBeInTheDocument();

  const completeBtn = queryByText("Mark Complete");

  expect(newTask).toHaveClass("Todo-task");
  expect(newTask).not.toHaveClass("Todo-complete");
  
  fireEvent.click(completeBtn);

  expect(newTask).toHaveClass("Todo-task Todo-complete");
});