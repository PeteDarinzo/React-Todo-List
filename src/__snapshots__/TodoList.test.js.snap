// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the todo snapshot 1`] = `
<DocumentFragment>
  <div>
    <h2>
      TODO List
    </h2>
    <form
      class="NewTodoForm"
    >
      <label
        class="NewTodoForm-label"
        for="task"
      >
        New Task
      </label>
      <input
        id="task"
        name="task"
        placeholder="Enter new task"
        type="text"
        value=""
      />
      <button
        class="NewTodoForm-button"
      >
        Submit
      </button>
    </form>
    <div
      class="ToDoList"
    />
  </div>
</DocumentFragment>
`;
